File: packages/sofinco-core/src/main/java/ch/sofinco/core/bridge/RepresentativeExampleBridge.java
package ch.sofinco.core.bridge;

import org.jahia.services.content.JCRNodeWrapper;

import java.util.Map;

/**
 * Bridge OSGi consommé par le mapping TypeScript de {@code packages/template-set}.
 *
 * <p>Côté TS, le service est récupéré par :
 * <pre>{@code
 *   server.osgi.getService("ch.sofinco.core.bridge.RepresentativeExampleBridge")
 * }</pre>
 *
 * <p>Le bridge fait l'adaptation entre le domaine métier Java (record
 * {@code RepresentativeExample} typé fortement) et le format consommable par JavaScript
 * (Map de clés/valeurs primitives).
 *
 * <p>Il lit lui-même les propriétés <b>à plat</b> sur le node reçu — le TS reste simple
 * ({@code bridge.getExample(node)}) sans gérer la traversée JCR.
 *
 * <h2>Deux origines possibles pour les paramètres</h2>
 * <p>Le node accepté peut être :
 * <ul>
 *   <li>une <b>PAGE</b> portant {@code sofmix:simulationParams} → {@code simProduct},
 *       {@code simSourceId}, {@code simAmount}, {@code simDuration}, {@code simScaleCode}.
 *       C'est la forme cible : un exemple représentatif décrit une offre sur une page, pas un
 *       bloc, et les valeurs deviennent consommables par tous les composants de la page ;</li>
 *   <li>le node du composant {@code sofnt:representativeExample} → {@code product},
 *       {@code sourceId} (mixin {@code sofmix:simulatorCta}), {@code amount}, {@code dueNumber},
 *       {@code scaleCode} (natifs). Forme héritée, conservée le temps que la migration
 *       {@code migrate-simulation-params-to-page.groovy} soit rejouée et vérifiée.</li>
 * </ul>
 * <p>Les deux jeux de noms ne coexistent jamais sur un même node : celui qui répond gagne.
 */
public interface RepresentativeExampleBridge {

    /**
     * Calcule et retourne l'exemple représentatif pour le node donné.
     *
     * @param componentNode une PAGE portant {@code sofmix:simulationParams}, ou le node d'un
     *                      composant {@code sofnt:representativeExample} (cf. javadoc de classe)
     * @return une Map consommable côté TS, ou {@code null} si l'exemple n'a pas pu être calculé
     *         (paramètres incomplets, config absente, APIM indisponible…)
     */
    Map<String, Object> getExample(JCRNodeWrapper componentNode);
}

File: packages/template-set/src/components/RepresentativeExample/representativeExampleBridge.ts
import { server } from "@jahia/javascript-modules-library";
import type { JCRNodeWrapper } from "org.jahia.services.content";
import { readString, readBoolean, type JavaRecord } from "#lib/javaBridge";

/**
 * Bridge OSGi Java pour le composant `sofnt:representativeExample`.
 *
 * Service Java : `ch.sofinco.core.bridge.RepresentativeExampleBridge`.
 * Méthode : `getExample(componentNode)` — lit directement les propriétés du
 * mixin `sofmix:simulatorCta` sur le node (simSourceId, simAmount, simDuration,
 * simPredefinedCreditType, scaleCode) et retourne le `JavaRecord` calculé.
 */

export const REPRESENTATIVE_EXAMPLE_BRIDGE = "ch.sofinco.core.bridge.RepresentativeExampleBridge";

export interface RepresentativeExampleBridge {
	getExample(componentNode: JCRNodeWrapper): JavaRecord | null;
}

export function getRepresentativeExampleBridge(): RepresentativeExampleBridge | null {
	const svc = server.osgi.getService(
		REPRESENTATIVE_EXAMPLE_BRIDGE,
	) as RepresentativeExampleBridge | null;
	if (!svc || typeof svc.getExample !== "function") return null;
	return svc;
}

/* ──────────────────────────────────────────────────────────────────────────
   Types & helpers de conversion JavaRecord → données typées
   (identiques à V1 — le shape Java est le même, seule la source change)
   ────────────────────────────────────────────────────────────────────────── */

export type ProductVariant = "pretPerso" | "creditRenouvelable" | "rachatCredit";

export interface ExampleRow {
	labelKey: string;
	value: string;
	highlighted: boolean;
	labelParam?: string;
}

export interface ExampleInsurance {
	monthlyAmount?: string;
	taea?: string;
	totalInsuranceCost?: string;
	monthlyWithInsurance?: string;
	lastWithInsurance?: string;
	dueNumber?: string;
	dueNumberMinusOne?: string;
}

export interface ExampleData {
	variant: ProductVariant;
	exampleAmount: string;
	rows: ExampleRow[];
	insurance?: ExampleInsurance;
	/**
	 * Texte d'assurance richtext optionnel défini sur le node config JCR
	 * (insurancePB / insuranceCR / insuranceRAC selon le variant).
	 * Si vide/null, le mapping retombe sur le fallback fr.json.
	 * Le `mention` du composant lui-même reste prioritaire sur cette valeur.
	 */
	insuranceTextOverride?: string;
}

export function toExampleData(record: JavaRecord | null): ExampleData | null {
	if (!record) return null;

	const rawRows = record["rows"] as ArrayLike<JavaRecord> | undefined;
	const rows: ExampleRow[] = [];
	if (rawRows) {
		for (let i = 0; i < rawRows.length; i++) {
			const r = rawRows[i];
			const labelParam = readString(r, "labelParam");
			rows.push({
				labelKey: readString(r, "labelKey"),
				value: readString(r, "value"),
				highlighted: readBoolean(r, "highlighted"),
				labelParam: labelParam || undefined,
			});
		}
	}

	const rawIns = record["insurance"] as JavaRecord | undefined;
	let insurance: ExampleInsurance | undefined;
	if (rawIns) {
		insurance = {
			monthlyAmount: readString(rawIns, "monthlyAmount") || undefined,
			taea: readString(rawIns, "taea") || undefined,
			totalInsuranceCost: readString(rawIns, "totalInsuranceCost") || undefined,
			monthlyWithInsurance: readString(rawIns, "monthlyWithInsurance") || undefined,
			lastWithInsurance: readString(rawIns, "lastWithInsurance") || undefined,
			dueNumber: readString(rawIns, "dueNumber") || undefined,
			dueNumberMinusOne: readString(rawIns, "dueNumberMinusOne") || undefined,
		};
	}

	const variantStr = readString(record, "variant");
	const variant: ProductVariant =
		variantStr === "pretPerso"
			? "pretPerso"
			: variantStr === "rachatCredit"
				? "rachatCredit"
				: "creditRenouvelable";

	const insuranceTextOverride = readString(record, "insuranceTextOverride");

	return {
		variant,
		exampleAmount: readString(record, "exampleAmount"),
		rows,
		insurance,
		insuranceTextOverride: insuranceTextOverride || undefined,
	};
}
